sql: deal with computed and default column generation in index backfi…

ajwerner opened the pull request:
This commit was mostly motivated by work on the new schema changer to enable
the behavior described in #47989, however, it also turns out to be a
prerequisite of the work to use virtual computed columns in secondary indexes.
Given we haven't released virtual computed columns, I'm going to omit a
release not for this PR.

The basic idea is that we need to track dependencies for computed columns
to make sure they are retrieved. Default expressions need to be evaluated
first. Much of the code is testing.

Release note: None

Commit:
sql: deal with computed and default column generation in index backfiller

This commit was mostly motivated by work on the new schema changer to enable
the behavior described in #47989, however, it also turns out to be a
prerequisite of the work to use virtual computed columns in secondary indexes.
Given we haven't released virtual computed columns, I'm going to omit a
release not for this PR.

The basic idea is that we need to track dependencies for computed columns
to make sure they are retrieved. Default expressions need to be evaluated
first. Much of the code is testing.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58990)
<!-- Reviewable:end -->

ajwerner commented:
I spent a bit of time firstly working on a low-level test to try to force the index backfill to happen and then also to do the backfill on a new column but that wasn't supported. Hence the small amount of testing here. This code gets exercised in the new schema changer code. Curious to hear takes.

ajwerner commented:
:( this code doesn't affect the virtual index code at all. I think this *just works* because of the way that the table fetcher works. In principle what this code enables is:

```
CREATE TABLE t (i INT PRIMARY KEY);
BEGIN;
    ALTER TABLE t ADD COLUMN j INT AS (i+1) VIRTUAL;
    CREATE INDEX idx ON t(j);
COMMIT;
```
But the above isn't supported for other reasons. I think what this means is this commit has zero testing. I think the test will need to synthesize some table state that isn't currently possible in order to exercise this code. I started typing it, I can finish it. It may come in handy one day.

thoszhang reviewed (commented):
I haven't looked at the code closely yet. I'm still trying to understand this example with the virtual column and the index added in the same transaction. If the fetcher used by the current index backfiller "just works" on already-existing public virtual columns (presumably because it just computes and returns values for the virtual columns like any other column?), and we currently have the ability to build indexes from non-public columns, then shouldn't your example also work once we get past `ADD COLUMN` being unimplemented for virtual columns?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-:-MQz_T8TD4w-btgQkXzU:b-mgmzi3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

ajwerner commented:
> I'm still trying to understand this example with the virtual column and the index added in the same transaction. If the fetcher used by the current index backfiller "just works" on already-existing public virtual columns (presumably because it just computes and returns values for the virtual columns like any other column?), and we currently have the ability to build indexes from non-public columns, then shouldn't your example also work once we get past ADD COLUMN being unimplemented for virtual columns?

Yes, I believe so. The missing logic there is that we need to change the add column schema change logic to know that it doesn't need to run a backfill on the primary index. I suppose I could just try to pick that up in order to test that this logic works.

lucy-zhang commented:
> I suppose I could just try to pick that up in order to test that this logic works.

"this logic" meaning what's on master, or with this PR? From what you're saying, it sounds like making the required changes to the add column logic on top of master, even without the changes to index backfill in this PR, will make https://github.com/cockroachdb/cockroach/pull/58990#issuecomment-759951489 work.

ajwerner commented:
> From what you're saying, it sounds like making the required changes to the add column logic on top of master, even without the changes to index backfill in this PR, will make #58990 (comment) work.

I don't think so. The table fetcher doesn't deal with computed expressions for columns that aren't public, hence the need for this change. Though, I guess, now that I say it, computed virtual columns could go straight to public, so then you're right.

lucy-zhang commented:
> The table fetcher doesn't deal with computed expressions for columns that aren't public, hence the need for this change.

Oh, that makes sense.

It seems like it'd be beneficial to backfill all indexes on virtual columns in the same way, regardless of whether the column is public. I'm not sure that we'd actually want to make virtual columns public immediately, since it's good to have all the columns become public simultaneously at the end of the schema change, and having (e.g.) `SELECT *` queries see some but not all columns being added during a schema change might cause new problems. In the new schema changer, I think we'd want to skip ahead to the backfilled state immediately, but not to the public state.

Could we modify the table fetcher to handle non-public virtual columns? It doesn't seem that ridiculous, given that the fetcher already handles non-public physical columns. We'd just need to guarantee that all the physical columns have already been backfilled, right? I don't have a good grasp on the details here.

lucy-zhang commented:
Another relevant possibility is something like this:

```
CREATE TABLE t ();
BEGIN;
    ALTER TABLE t ADD COLUMN i INT;
    ALTER TABLE t ADD COLUMN j INT AS (i+1) VIRTUAL;
    CREATE INDEX idx ON t(j);
COMMIT;
```

ajwerner commented:
> Could we modify the table fetcher to handle non-public virtual columns? It doesn't seem that ridiculous, given that the fetcher already handles non-public physical columns.

It does? I thought it was enforcing the whole thing about the column not being readable. 

——— 

That last possibility doesn’t work because we’ll do the column backfill for the primary index first.

lucy-zhang commented:
Summarizing some offline discussion: We'll start off in 21.1 with all added virtual columns becoming public immediately, and disallowing any other schema changes that would be incompatible with this (e.g., adding virtual columns on non-public columns, or adding constraints on virtual columns that need validation). Since adding secondary indexes on public virtual columns is already supported, there's nothing else to be done for that. So this change is actually addressing an entirely separate concern and is pretty specific to the new schema changer itself.

lucy-zhang commented:
> Since adding secondary indexes on public virtual columns is already supported

Actually, maybe not. The table fetcher seems to already handle this, but we need to make sure.

ajwerner commented:
I fixed this up to work with virtual computed columns. I also added testing. That was the harder part. Github seems to be struggling this morning but when it gets its act together the build should pass now.

ajwerner commented:
@adityamaru you should probably have a look at this as well.

thoszhang reviewed (commented):
This mostly LGTM, but I do worry about testing more cases. It seems like a pain to write these low-level tests, and maybe our best bet for more comprehensive testing is the random schema change tests (both for indexes on virtual computed columns and for the new index backfiller column changes).

> Reviewed 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-:-MS069cc5CrsXsA11ZO7:b-f5s55d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @RaduBerinde)

___
*[pkg/sql/indexbackfiller_test.go, line 194 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS014v04L05mivbwasZ:-MS014v04L05mivbwas_:b-w1vm55) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L194)):*
> ```Go
> }
> 
> type indexBackfillTestCase struct {
> ```

nit: I think this and `run` should go inside the test function. It is annoying but ultimately I think it's good to have these scoped to the test.
___
*[pkg/sql/indexbackfiller_test.go, line 203 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS01J3D9ajV20xEKhOf:-MS01J3D9ajV20xEKhOg:b-ejbci6) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L203)):*
> ```Go
> 
> 	// setupDesc should mutate the descriptor such that the mutation with
> 	// id 1.
> ```

nit: finish this sentence
___
*[pkg/sql/indexbackfiller_test.go, line 209 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS01KuO01iiXmjYhEkB:-MS01KuO01iiXmjYhEkC:bgqxdhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L209)):*
> ```Go
> }
> 
> func (test indexBackfillTestCase) run(t *testing.T) {
> ```

I think this function would benefit from a few more comments inline in general.
___
*[pkg/sql/indexbackfiller_test.go, line 391 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS01wZM4EYxAlRUY4ng:-MS01wZM4EYxAlRUY4nh:bdixr9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L391)):*
> ```Go
> 			mut.Mutations[len(mut.Mutations)-1].State = descpb.DescriptorMutation_DELETE_AND_WRITE_ONLY
> 
> 			// This tests that we can backfill new computed, stored columns into
> ```

I think this is a stray comment.
___
*[pkg/sql/indexbackfiller_test.go, line 500 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS04U8LElZJsKhodeNJ:-MS04U8MCRRzjhNjaOtu:b-b0xol) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L500)):*
> ```Go
> 	))
> 
> 	require.NoError(t, fetcher.StartScan(ctx, txn, spans, false, 0, true))
> ```

I think you'll have a merge conflict with #59208 which updated the fetcher scan arguments.
___
*[pkg/sql/backfill/backfill.go, line 409 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MRvZi_hC2oSwEPB9hQO:-MRvZi_hC2oSwEPB9hQP:b-b36p1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/backfill/backfill.go#L409)):*
> ```Go
> 	addedCols []descpb.ColumnDescriptor
> 
> 	// computedCols are the column in this index which are computed and are do
> ```

nit: `columns`
___
*[pkg/sql/backfill/backfill.go, line 701 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MRwzKESElhbTWr66n3M:-MRwzKESElhbTWr66n3N:b-dfn4gj) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/backfill/backfill.go#L701)):*
> ```Go
> 				isPrimaryIndex := idx.GetEncodingType(desc.GetPrimaryIndexID()) == descpb.PrimaryIndexEncoding
> 				if (idxContainsColumn || isPrimaryIndex) &&
> 					!(ib.cols[i].Virtual || i >= len(desc.Columns)) {
> ```

nit: I think I'd prefer `!ib.cols[i].Virtual && i < len(desc.Columns)`.

I also find this opaque since it relies on the ordering of `ib.cols`. Add a comment here explaining what's going on?
___
*[pkg/sql/catalog/schemaexpr/computed_column.go, line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MRx0OzMB-9LIit6BxUG:-MRx0OzMB-9LIit6BxUH:b-xh1u7n) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/catalog/schemaexpr/computed_column.go#L172)):*
> ```Go
> // slice of input column descriptors, or nil if none of the input column
> // descriptors have computed expressions. The caller provides the set of
> // valid to which the expr may refer.
> ```

nit: `valid columns`
___
*[pkg/sql/catalog/schemaexpr/computed_column.go, line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MRx1Rw3CmgZHfFI3jAK:-MRx1Rw3CmgZHfFI3jAL:b-jldtyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/catalog/schemaexpr/computed_column.go#L179)):*
> ```Go
> //
> // Note that the order of columns is critical. Expressions cannot reference
> // columns that come after them in cols.
> ```

nit: Update this comment with the new variable names


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Are there other cases you'd like to see tested? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-:-MS6x0M56zfXayFHLP3i:bnigxl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @lucy-zhang, and @RaduBerinde)

___
*[pkg/sql/indexbackfiller_test.go, line 194 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS014v04L05mivbwasZ:-MS6twhe9zT4dfjxdYQL:b-ub4156) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L194)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: I think this and `run` should go inside the test function. It is annoying but ultimately I think it's good to have these scoped to the test.
</blockquote></details>

sure, that's where it started. 
___
*[pkg/sql/indexbackfiller_test.go, line 203 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS01J3D9ajV20xEKhOf:-MS6v8KYATRNc90ZeUyR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L203)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: finish this sentence
</blockquote></details>

Done.
___
*[pkg/sql/indexbackfiller_test.go, line 209 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS01KuO01iiXmjYhEkB:-MS6wxhd6HRrlcLxI25F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L209)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think this function would benefit from a few more comments inline in general.
</blockquote></details>

Done.
___
*[pkg/sql/indexbackfiller_test.go, line 391 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS01wZM4EYxAlRUY4ng:-MS6w8Z7EUdhpmTnQeGv:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L391)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think this is a stray comment.
</blockquote></details>

Removed.
___
*[pkg/sql/indexbackfiller_test.go, line 500 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS04U8LElZJsKhodeNJ:-MS6vGiN3Ki98mMTHsBc:bnlamz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/indexbackfiller_test.go#L500)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

I think you'll have a merge conflict with #59208 which updated the fetcher scan arguments.
</blockquote></details>

resolved.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
@adityamaru any chance I could get you to sign off on this guy?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-:-MS8WHl1EXr4yJUhbq1H:b4ow3np)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @lucy-zhang, and @RaduBerinde)



<!-- Sent from Reviewable.io -->

adityamaru commented:
> @adityamaru any chance I could get you to sign off on this guy?
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-:-MS8WHl1EXr4yJUhbq1H:b4ow3np)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @lucy-zhang, and @RaduBerinde)

sorry, giving it a look right now

adityamaru commented:
LGTM! Those unit tests are really cool.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58990)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @lucy-zhang)

___
*[pkg/sql/indexbackfiller_test.go, line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS8zjPK7261AmirUnhG:-MS8zjPK7261AmirUnhH:bmwuetm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc5e1b5fcc892ae4e923651b798e19015107f95/pkg/sql/indexbackfiller_test.go#L349)):*
> ```Go
> 	}
> 
> 	// fetchIndex fetches the contents of an a table index returning the results
> ```

I think there would be a lot of value in exposing something like this in SQL via a special statement or a virtual table. It would output a representation of the KVs that we could check for in various tests (including logic tests).


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58990)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @RaduBerinde)

___
*[pkg/sql/indexbackfiller_test.go, line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS8zjPK7261AmirUnhG:-MS99CaeCm5LQjl8v_W0:b-70ye3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc5e1b5fcc892ae4e923651b798e19015107f95/pkg/sql/indexbackfiller_test.go#L349)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think there would be a lot of value in exposing something like this in SQL via a special statement or a virtual table. It would output a representation of the KVs that we could check for in various tests (including logic tests).
</blockquote></details>

I agree. I think I'd like to defer that to a later time but I've filed an issue. https://github.com/cockroachdb/cockroach/issues/59549 
___
*[pkg/sql/backfill/backfill.go, line 409 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MRvZi_hC2oSwEPB9hQO:-MS9DLjYBA6QWNxIJZ3j:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/backfill/backfill.go#L409)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: `columns`
</blockquote></details>

Done.
___
*[pkg/sql/backfill/backfill.go, line 701 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MRwzKESElhbTWr66n3M:-MS9DL5r1i7nT63PxFP2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/backfill/backfill.go#L701)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: I think I'd prefer `!ib.cols[i].Virtual && i < len(desc.Columns)`.

I also find this opaque since it relies on the ordering of `ib.cols`. Add a comment here explaining what's going on?
</blockquote></details>

Done.
___
*[pkg/sql/catalog/schemaexpr/computed_column.go, line 172 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MRx0OzMB-9LIit6BxUG:-MS9DhPb4axgs075ZHIk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/catalog/schemaexpr/computed_column.go#L172)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: `valid columns`
</blockquote></details>

Done.
___
*[pkg/sql/catalog/schemaexpr/computed_column.go, line 179 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MRx1Rw3CmgZHfFI3jAK:-MS9DexW0Bf6g40fZnlu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59b52c7754f23fb6c70998b43f5fa4689790962e/pkg/sql/catalog/schemaexpr/computed_column.go#L179)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

nit: Update this comment with the new variable names
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58990)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @lucy-zhang)

___
*[pkg/sql/indexbackfiller_test.go, line 349 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58990#-MS8zjPK7261AmirUnhG:-MS9FfIS4a301-V5w3A0:bk77kjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebc5e1b5fcc892ae4e923651b798e19015107f95/pkg/sql/indexbackfiller_test.go#L349)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I agree. I think I'd like to defer that to a later time but I've filed an issue. https://github.com/cockroachdb/cockroach/issues/59549 
</blockquote></details>

Thanks, yeah I didn't mean in this PR :)


<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+
